### PR TITLE
Meal entry

### PR DESCRIPTION
**Pull request by rhantak**
- Add functionality for user to add food by upc code
- Add functionality for user to build dishes out of foods
- Add functionality for user to build meals out of dishes
- Add functionality for user to log meals to their account
- Meals no longer validate the presence of a gut feeling, since they will be added later by the user
- Any user specific page now requires a visitor to log in

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -6,4 +6,8 @@ class ApplicationController < ActionController::Base
   def current_user
     @current_user ||= User.find(session[:user_id]) if session[:user_id]
   end
+
+  def require_login
+    render file: "/public/404" if current_user.nil?
+  end
```
Comment (johnktravers) [inline app/controllers/application_controller.rb]: Good thinking with the restriction to logged in users

File: app/models/meal.rb [hunk 2]
```diff
@@ -1,16 +1,26 @@
...
+
+  def create_meal_ingredients
+    dishes.each do |dish|
+      dish.foods.each do |food|
+        food.ingredients.each do |ingredient|
+          MealIngredient.create(meal_id: id, ingredient_id: ingredient.id)
+        end
+      end
+    end
+  end
```
Comment (johnktravers) [inline app/models/meal.rb]: I really like that you moved these into the model from the controller action

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -74,4 +74,29 @@
...
+        'refresh_token' => 'another_token',
+        'expires_at' => 1_354_920_555,
+        'expires' => true
+      }
+    )
+
+    Rails
+      .application
+      .env_config['omniauth.auth'] = OmniAuth.config.mock_auth[:google_oauth2]
+  end
```
Comment (johnktravers) [inline spec/rails_helper.rb]: Really like this helper method. Good work!

File: spec/models/food_spec.rb [hunk 4]
```diff
@@ -18,4 +18,40 @@
...
+            "description": "CLASSICO, TOMATO & BASIL PASTA SAUCE, TOMATO & BASIL, TOMATO & BASIL",
+            "dataType": "Branded",
+            "gtinUpc": "041129077122",
+            "publishedDate": "2019-12-06",
+            "brandOwner": "New World Pasta Company",
+            "ingredients": "TOMATO PUREE (WATER, TOMATO PASTE), DICED TOMATOES IN JUICE (TOMATOES, TOMATO JUICE, CITRIC ACID, CALCIUM CHLORIDE), CONTAINS 2% OR LESS OF: OLIVE OIL, ONIONS, SALT, BASIL, GARLIC, SPICES, NATURAL FLAVOR.",
+            "allHighlightFields": "<b>GTIN/UPC</b>: <em>041129077122</em>",
+            "score": -747.62915
+        }
+      expect(food.ingredient_list(food_info)).to eq( ["TOMATO PUREE", "DICED TOMATOES IN JUICE"] )
```
Comment (Jonpatt92) [inline spec/models/food_spec.rb]: Good test and great functionality

**Pull request merged**